File: scala/leet_code/623.scala
/** This is the Solution object. It contains a method to add a row of nodes with a given value at a given depth in a binary tree.
  * The method returns the root of the modified tree.
  *
  * @example
  * val root = new TreeNode(4, new TreeNode(2, new TreeNode(3), new TreeNode(1)), new TreeNode(6, new TreeNode(5)))
  * val result = Solution.addOneRow(root, 1, 2)
  * // result: TreeNode(4, TreeNode(1, TreeNode(2, TreeNode(3), TreeNode(1))), TreeNode(1, TreeNode(6, TreeNode(5))))
  *
  * @note The method modifies the input tree by adding a row of nodes at the specified depth.
  */
object Solution {

  /** Adds a row of nodes with a given value at a given depth in a binary tree.
    *
    * @param root The root of the binary tree.
    * @param `val` The value of the nodes to be added.
    * @param depth The depth at which the nodes should be added.
    * @return The root of the modified tree.
    * @throws IllegalArgumentException if the depth is less than 1.
    */
  def addOneRow(root: TreeNode, `val`: Int, depth: Int): TreeNode = {
    if (depth < 1) {
      throw new IllegalArgumentException("Depth must be greater than 0")
    }
    helper(root, `val`, depth, 1)
  }

  /** Helper function to recursively add nodes at a given depth in a binary tree.
    *
    * @param root The root of the binary tree.
    * @param `val` The value of the nodes to be added.
    * @param depth The depth at which the nodes should be added.
    * @param layer The current layer of the tree.
    * @return The root of the modified tree.
    */
  private def helper(
      root: TreeNode,
      `val`: Int,
      depth: Int,
      layer: Int
  ): TreeNode = {
    if (root == null) return null
    if (depth == 1) {
      val newNode = new TreeNode(_value = `val`, _left = root)
      return newNode
    }
    if (layer == depth - 1) {
      var newNode = new TreeNode(_value = `val`, _left = root.left)
      root.left = newNode
      newNode = new TreeNode(_value = `val`, _right = root.right)
      root.right = newNode
    } else {
      root.left = helper(root.left, `val`, depth, layer + 1)
      root.right = helper(root.right, `val`, depth, layer + 1)
    }
    root
  }
}

File: csharp/leet_code/623.cs
/// <summary>
/// Definition for a binary tree node.
/// </summary>
public class TreeNode
{
    public int val;
    public TreeNode left;
    public TreeNode right;

    /// <summary>
    /// Constructor for a binary tree node.
    /// </summary>
    /// <param name="val">The value of the node.</param>
    /// <param name="left">The left child of the node.</param>
    /// <param name="right">The right child of the node.</param>
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/// <summary>
/// This class provides a solution for adding a row to a binary tree at a given depth.
/// </summary>
public class Solution
{
    /// <summary>
    /// This method adds a row to the binary tree at the given depth.
    /// </summary>
    /// <param name="root">The root of the binary tree.</param>
    /// <param name="val">The value to be added to the new row.</param>
    /// <param name="depth">The depth at which the new row should be added.</param>
    /// <returns>The root of the binary tree after adding the new row.</returns>
    public TreeNode AddOneRow(TreeNode root, int val, int depth)
    {
        if (depth == 1)
            return new TreeNode(val, root, null);

        AddOneRow(root, val, depth, 1);
        return root;
    }

    /// <summary>
    /// This method is a helper for AddOneRow. It recursively adds a row to the binary tree at the given depth.
    /// </summary>
    /// <param name="root">The root of the binary tree.</param>
    /// <param name="val">The value to be added to the new row.</param>
    /// <param name="depth">The depth at which the new row should be added.</param>
    /// <param name="currDepth">The current depth in the recursion.</param>
    private void AddOneRow(TreeNode root, int val, int depth, int currDepth)
    {
        if (currDepth + 1 == depth)
        {
            var left = root.left;
            var right = root.right;

            var newNodeLeft = new TreeNode(val, root.left, null);
            var newNodeRight = new TreeNode(val, null, root.right);

            root.left = newNodeLeft;
            root.right = newNodeRight;
        }
        else
        {
            if (root.left is not null)
                AddOneRow(root.left, val, depth, currDepth + 1);

            if (root.right is not null)
                AddOneRow(root.right, val, depth, currDepth + 1);
        }
    }
}
